finality-aleph/src/data_io.rs

```diff
@@ -359,82 +1001,228 @@ where
+        let mut prev_best_block: Option<BlockHashNum<B>> = None;
+        loop {
+            let delay = futures_timer::Delay::new(Duration::from_millis(REFRESH_INTERVAL));
+            tokio::select! {
+                _ = delay => {
+                    let new_best_header = self.get_best_header().await;
+                    let new_best_block = BlockHashNum::<B>::new(new_best_header.hash(), *new_best_header.number());
+                    if new_best_header.number() <= &self.max_block_num {
+                        self.best_block_in_session = Some(new_best_block.clone());
+                    } else if prev_best_block.is_none() || prev_best_block.as_ref().unwrap().num < self.max_block_num {
+                            // This is either the first iteration in the loop, or the case when the `new_best_header` first
+                            // exceeds `max_block_num`. We don't use any tricks just reduce the `new_best_header` to `max_block_num`
+                            let reduced_header = reduce_header_to_num(self.client.clone(), new_best_header.clone(), self.max_block_num);
+                            self.best_block_in_session = Some(BlockHashNum::<B>::new(reduced_header.hash(), *reduced_header.number()));
+                        } else {
+                            // Here both the `prev_best_block` and `new_best_header` are above `max_block_num`, we optimize.
+                            let prev = prev_best_block.unwrap();
+                            let reduced_header = reduce_header_to_num(self.client.clone(), new_best_header.clone(), prev.num);
+                            if reduced_header.hash() != prev.hash {
+                                // The new_best_block is not a descendant of prev_best_block, we need to update.
+                                // In the opposite case we do nothing, as the `best_block_in_session` is already correct.
+                                let reduced_header = reduce_header_to_num(self.client.clone(), new_best_header.clone(), self.max_block_num);
+                                self.best_block_in_session = Some(BlockHashNum::<B>::new(reduced_header.hash(), *reduced_header.number()));
+                            }
+                        }
+
+                    self.update_data();
+                    prev_best_block = Some(new_best_block);
+                }
+                _ = &mut exit => {
+                    debug!(target: "afa", "Task for refreshing best chain received exit signal. Terminating.");
+                    return;
+                }
+            }
         }
     }
 }
 
 #[async_trait]
-impl<B: BlockT> aleph_bft::DataProvider<AlephDataFor<B>> for DataProvider<B> {
-    async fn get_data(&mut self) -> AlephDataFor<B> {
-        let best = *self.proposed_block.lock();
+impl<B: BlockT> aleph_bft::DataProvider<AlephData<B>> for DataProvider<B> {
+    async fn get_data(&mut self) -> AlephData<B> {
+        let data: AlephData<B> = self.data_to_propose.lock().clone();
```

Comment:
Huh, this is typed? Is it because it could technically want to `clone` the wrapping lock struct? Then I think explicitly calling `deref` is clearer.

Reply:
Changed to `let data: AlephData<B> = (*self.data_to_propose.lock()).clone();` 
Is this what you propose?

Reply:
I wanted to get rid of the type annotation mostly. The `*` should work for that, doesn't it?

Reply:
You mean just remove the type annotation here? Get rid of clone?

Reply:
Remove the type annotation – my intuition is that if you had to write one then something likely went wrong, especially in a case in which it should be obvious to the compiler. I think the compiler should no longer complain with the explicit `*`, but I'm not sure.

I don't think you can/want to get rid of `clone` here, right?

Reply:
Ok, sure, annotation is gone, clone must stay because no `Copy`.